common/junit-platform-native/src/main/java/org/graalvm/junit/platform/JUnitPlatformFeature.java

```diff
@@ -93,6 +96,46 @@ public void beforeAnalysis(BeforeAnalysisAccess access) {
         Launcher launcher = LauncherFactory.create();
         TestPlan testplan = discoverTestsAndRegisterTestClassesForReflection(launcher, selectors);
         ImageSingletons.add(NativeImageJUnitLauncher.class, new NativeImageJUnitLauncher(launcher, testplan));
+
+        Method registerAllDeclaredMethods = findMethodOrNull(RuntimeReflection.class, "registerAllDeclaredMethods", Class.class);
+        if (registerAllDeclaredMethods != null) {
+            // graalvm-23.0+ with `RuntimeReflection#registerAllDeclaredMethods` method.
+            ClassLoader applicationLoader = access.getApplicationClassLoader();
+            Class<?> typeSafeMatcher = findClassOrNull(applicationLoader, "org.hamcrest.TypeSafeMatcher");
+            Class<?> typeSafeDiagnosingMatcher = findClassOrNull(applicationLoader, "org.hamcrest.TypeSafeDiagnosingMatcher");
+            if (typeSafeMatcher != null || typeSafeDiagnosingMatcher != null) {
+                BiConsumer<DuringAnalysisAccess, Class<?>> registerMatcherForReflection = (a, c) -> {
+                    try {
+                        registerAllDeclaredMethods.invoke(null, c);
+                    } catch (ReflectiveOperationException e) {
+                        throw new RuntimeException(e);
+                    }
+                };
+                if (typeSafeMatcher != null) {
+                    access.registerSubtypeReachabilityHandler(registerMatcherForReflection, typeSafeMatcher);
```

Comment:
How do we guarantee this does not affect user code by registering their classes?

Reply:
@vjovanov: Good question. I believe it's acceptable to register even the user classes. This is because subclasses of `org.hamcrest.TypeSafeMatcher` and `org.hamcrest.TypeSafeDiagnosingMatcher` that utilize the default constructor rely on `Class.getDeclaredMethods()` to function properly. You can refer to [TypeSafeMatcher#TypeSafeMatcher()](https://github.com/hamcrest/JavaHamcrest/blob/master/hamcrest/src/main/java/org/hamcrest/TypeSafeMatcher.java#L25), which indirectly invokes [ReflectiveTypeFinder#findExpectedType()](https://github.com/hamcrest/JavaHamcrest/blob/master/hamcrest/src/main/java/org/hamcrest/internal/ReflectiveTypeFinder.java#L43). Ideally, we would only mark the `matchesSafely` method for `getDeclaredMethods`, but unfortunately, this is not possible.